cpp/src/hash/sha512_hash.cu

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sha_hash.cuh"
+
+#include <cudf/column/column.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace hashing {
+namespace detail {
+
+namespace {
+
+// Need alignas(16) to avoid compiler bug.
+struct alignas(16) sha512_hash_state {
+  uint64_t message_length = 0;
+  uint32_t buffer_length  = 0;
+  uint64_t hash_value[8]  = {0x6a09e667f3bcc908,
+                             0xbb67ae8584caa73b,
+                             0x3c6ef372fe94f82b,
+                             0xa54ff53a5f1d36f1,
+                             0x510e527fade682d1,
+                             0x9b05688c2b3e6c1f,
+                             0x1f83d9abfb41bd6b,
+                             0x5be0cd19137e2179};
+  uint8_t buffer[128];
+};
+
+struct SHA512Hash : HashBase<SHA512Hash> {
+  __device__ inline SHA512Hash(char* result_location) : HashBase<SHA512Hash>(result_location) {}
+
+  // Intermediate data type storing the hash state
+  using hash_state = sha512_hash_state;
+  // The word type used by this hash function
+  using sha_word_type = uint64_t;
+  // Number of bytes processed in each hash step
+  static constexpr uint32_t message_chunk_size = 128;
+  // Digest size in bytes
+  static constexpr uint32_t digest_size = 128;
+  // Number of bytes used for the message length
+  static constexpr uint32_t message_length_size = 16;
+
+  void __device__ inline hash_step(hash_state& state) { sha512_hash_step(state); }
+
+  hash_state state;
+};
+
+}  // namespace
+
+std::unique_ptr<column> sha512(table_view const& input,
+                               rmm::cuda_stream_view stream,
+                               rmm::mr::device_memory_resource* mr)
+{
+  string_scalar const empty_result(
+    "cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec"
+    "2f63b931bd47417a81a538327af927da3e");
```

Comment:
What's the significance of these hardcoded empty results?  Are these some kind of standard value or randomly chosen? Why wouldn't an empty string be more appropriate?

Reply:
These are the hash values you get for zero-length inputs. This is hardcoded as a way to handle the case where there are zero columns in the table. I'm not sure if the logic would properly handle zero-length inputs, but I will check that and simplify it if possible.

Reply:
I was able to refactor and remove this in fdb7c8fd17a35db978b9d2711d5ada33961d47b4. Resolving.